### PR TITLE
Add pool instantiation strategy

### PR DESCRIPTION
**Pull request by ajreynol**
Adds an instantiation strategy based on user-provided pool annotations.

The next PR will connect this to quantifiers engine.

### CODE REVIEW

File: src/theory/quantifiers/term_tuple_enumerator.cpp [hunk 1]
```diff
@@ -490,6 +491,48 @@ Node TermTupleEnumeratorBasic::getTerm(size_t variableIx, size_t term_index)
...
+
+ protected:
+  /** Pointer to the term pool utility */
+  TermPools* d_tp;
+  /** The pool annotation */
+  Node d_pool;
+  /**  a list of terms for each id */
+  std::map<size_t, std::vector<Node> > d_poolList;
+  /** prepare terms gets the terms from the pool */
+  virtual size_t prepareTerms(size_t variableIx) override
```
Comment (HanielB) [inline src/theory/quantifiers/term_tuple_enumerator.cpp:519]: you need to use `virtual` here?

File: src/theory/quantifiers/term_tuple_enumerator.cpp [hunk 2]
```diff
@@ -490,6 +491,48 @@ Node TermTupleEnumeratorBasic::getTerm(size_t variableIx, size_t term_index)
...
+  virtual ~TermTupleEnumeratorPool() = default;
+
+ protected:
+  /** Pointer to the term pool utility */
+  TermPools* d_tp;
+  /** The pool annotation */
+  Node d_pool;
+  /**  a list of terms for each id */
+  std::map<size_t, std::vector<Node> > d_poolList;
+  /** prepare terms gets the terms from the pool */
```
Comment (HanielB) [inline src/theory/quantifiers/term_tuple_enumerator.cpp:518]:
```suggestion
  /** gets the terms from the pool */
```

File: src/theory/quantifiers/inst_strategy_pool.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ ** in the top-level source directory and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Enumerative instantiation
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_pool.h:12]: Update

File: src/theory/quantifiers/inst_strategy_pool.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace cvc5 {
+namespace theory {
+namespace quantifiers {
+
+/**
+ * Pool-based instantiation. This implements a strategy for instantiating
+ * quantifiers based on user-provided pool annotations.
+ *
+ * When check is invoked, this strategy instantiates considers each
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_pool.h:30]:
```suggestion
 * When check is invoked, this strategy considers each
```

File: src/theory/quantifiers/inst_strategy_pool.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ * INST_POOL annotation on quantified formulas. For each annotation, it
+ * instantiates the associated quantified formula with the Cartesian
+ * product of terms currently in the pool, using efficient techniques for
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_pool.h:33]: give examples of these efficient techniques

File: src/theory/quantifiers/inst_strategy_pool.cpp [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        d_userPools[q].push_back(p);
+      }
+    }
+  }
+}
+
+void InstStrategyPool::check(Theory::Effort e, QEffort quant_e)
+{
+  bool doCheck = !d_userPools.empty();
+  if (!doCheck)
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_pool.cpp:69]:
```suggestion
    if (d_userPools.empty())
```

File: src/theory/quantifiers/term_tuple_enumerator.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/term_tuple_enumerator.cpp:520]: No

File: src/theory/quantifiers/term_tuple_enumerator.cpp [hunk 7]
```diff
@@ -491,6 +492,48 @@ Node TermTupleEnumeratorBasic::getTerm(size_t variableIx, size_t term_index)
...
+
+ protected:
+  /** Pointer to the term pool utility */
+  TermPools* d_tp;
+  /** The pool annotation */
+  Node d_pool;
+  /**  a list of terms for each id */
+  std::map<size_t, std::vector<Node> > d_poolList;
+  /** gets the terms from the pool */
+  virtual size_t prepareTerms(size_t variableIx) override
```
Comment (HanielB) [inline src/theory/quantifiers/term_tuple_enumerator.cpp:520]:
```suggestion
  size_t prepareTerms(size_t variableIx) override
```

File: src/theory/quantifiers/term_tuple_enumerator.cpp [hunk 8]
```diff
@@ -491,6 +492,48 @@ Node TermTupleEnumeratorBasic::getTerm(size_t variableIx, size_t term_index)
...
+  {
+    Assert(d_pool.getNumChildren() > variableIx);
+    // prepare terms from pool
+    d_poolList[variableIx].clear();
+    d_tp->getTermsForPool(d_pool[variableIx], d_poolList[variableIx]);
+    Trace("pool-inst") << "Instantiation Terms for child " << variableIx << ": "
+                       << d_poolList[variableIx] << std::endl;
+    return d_poolList[variableIx].size();
+  }
+  virtual Node getTerm(size_t variableIx, size_t term_index) override
```
Comment (HanielB) [inline src/theory/quantifiers/term_tuple_enumerator.cpp:530]:
```suggestion
  Node getTerm(size_t variableIx, size_t term_index) override
```

**Review by HanielB — APPROVED** [review]
LGTM, just minor suggestions

**Pull request merged**